File: 백준/Gold/17136. 색종이 붙이기/색종이 붙이기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	static StringBuilder sb = new StringBuilder();
	static int result;
	static int[][] arr = new int[10][10];
	static int[] paper = {0, 5, 5, 5, 5, 5 };

	public static void main(String[] args) throws IOException {
		for (int i = 0; i < 10; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < 10; j++) {
				arr[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		

		result = Integer.MAX_VALUE;
		dfs(0, 0, 0);
		
		if(result == Integer.MAX_VALUE) {
			System.out.println(-1);
		}else {
			
			System.out.println(result);
		}
	}

	static void dfs(int x, int y, int cnt) {
		if (x >= 9 && y > 9) {
			result = Math.min(result, cnt);
			return;
		}
		
		if(result <= cnt) return;


		if (y > 9) {
			dfs(x + 1, 0, cnt);
			return;
		}

		if (arr[x][y] == 1) {
			for (int i = 1; i <= 5 ; i++) {
				if (paper[i] > 0 && possible(x, y, i)) {
					attach(x, y, i, 0);
					paper[i]--;
					dfs(x, y + 1, cnt + 1);
					attach(x, y, i, 1);
					paper[i]++;
				}
			}
		} else {
			dfs(x, y + 1, cnt);
		}

	}

	static void attach(int x, int y, int size, int num) {
		for (int i = x; i < x + size; i++) {
			for (int j = y; j < y + size; j++) {
				arr[i][j] = num;

			}
		}
	}

	static boolean possible(int x, int y, int size) {
		for (int i = x; i < x + size; i++) {
			for (int j = y; j < y + size; j++) {
				if (!check(i, j))
					return false;
				
				if(arr[i][j] == 0) return false;
			}
		}

		return true;
	}

	static boolean check(int x, int y) {
		return x >= 0 && x < 10 && y >= 0 && y < 10;
	}

}

File: 백준/Gold/17136. 색종이 붙이기/색종이 붙이기.py
def dfs(x, y, cnt):
    global result

    if x >= 9 and y > 9:
        result = min(result, cnt)
        return

    if result <= cnt:
        return

    if (y > 9):
        dfs(x + 1, 0, cnt)
        return

    if(arr[x][y] == 1):
        for i in range(1, 6):
            if(paper[i] > 0 and possible(x, y, i)):
                attach(x, y, i, 0)
                paper[i] -= 1
                dfs(x, y + 1, cnt + 1)
                attach(x, y, i, 1)  # 색종이 다시 떼기
                paper[i] += 1
    else:
        dfs(x, y + 1, cnt)

# 색종이를 붙일 수 있는지
def possible(x, y, size):
    for i in range(x, size + x):
        for j in range(y, size + y):
            if(not check(i, j) or arr[i][j] == 0): return False

    return True

# 색종이 붙이기
def attach(x, y, size, num):
    for i in range(x, size + x):
        for j in range(y, size + y):
            arr[i][j] = num


def check(ax, ay):
    return ax >= 0 and ax < 10 and ay >= 0 and ay < 10
arr = []
inf = int(1e9)
paper = [0, 5, 5, 5, 5, 5]
result = inf

for _ in range(10):
    arr.append(list(map(int, input().split())))


dfs(0, 0, 0)

print(-1 if result == inf else result)